esp-hal/src/spi/master.rs

```diff
@@ -1092,6 +1093,217 @@ mod dma {
         pub fn clear_interrupts(&mut self, interrupts: EnumSet<SpiInterrupt>) {
             self.spi.clear_interrupts(interrupts);
         }
+
+        fn is_done(&self) -> bool {
+            if self.tx_transfer_in_progress && !self.channel.tx.is_done() {
+                return false;
+            }
+            if self.spi.busy() {
+                return false;
+            }
+            if self.rx_transfer_in_progress {
+                // If this is an asymmetric transfer and the RX side is smaller, the RX channel
+                // will never be "done" as it won't have enough descriptors/buffer to receive
+                // the EOF bit from the SPI. So instead the RX channel will hit
+                // a "descriptor empty" which means the DMA is written as much
+                // of the received data as possible into the buffer and
+                // discarded the rest. The user doesn't care about this discarded data.
+
+                if !self.channel.rx.is_done() && !self.channel.rx.has_dscr_empty_error() {
+                    return false;
+                }
+            }
+            true
+        }
+
+        fn wait_for_idle(&mut self) {
+            while !self.is_done() {
+                // Wait for the SPI to become idle
+            }
+            self.rx_transfer_in_progress = false;
+            self.tx_transfer_in_progress = false;
+            fence(Ordering::Acquire);
+        }
+
+        async fn wait_for_idle_async(&mut self) {
+            // As a future enhancement, setup Spi Future in here as well.
+
+            /// An optional future.
+            enum OptionalFuture<T: Future> {
```

Comment:
Nice idea but the naming confused me a bit - maybe `FutureOrValue` ... `FutureOrReady` ... well probably my suggestions are not much better 🤷‍♂️

Reply:
It's probably simpler to just introduce a new future type in the DMA module that doesn't clear the interrupt bits on completion. (I was gonna add this eventually for the SpiTransfer)

Reply:
>  that doesn't clear the interrupt bits

And starts with those interrupt bits set in the first place, otherwise the first transfer will wait forever :)

Reply:
Why would it wait forever? The interrupt bit will eventually be set when the transfer is done, no?

Reply:
Well, no. Before the first transfer we're still waiting for the peripheral to be idle. If the idle condition is that the interrupt request bit is high, then we'll hang. We can be more clever about it, sure, but we're not at the moment.

Reply:
I've just reverted this change. Manually pending an interrupt bit and introducing a new pair of future types is actually a bigger change in terms of LOC, and we can do it separately - because I like the idea and wasting 2 bytes to track transfer states here is unnecessary.

If we rely on the hardware interrupt bits, we can even recover from cases where the user forgot the peripheral mid-transfer and re-created it using unasfe code. It's just weird that technically this would maybe be one step backwards from the DMA interrupt semantics change if I understand that correctly.

Reply:
> It's just weird that technically this would maybe be one step backwards from the DMA interrupt semantics change if I understand that correctly.

Yes, three steps forward and one step back. The clearing of the interrupt bit is the minor part of the semantics imo. The important part was treating the INT_ENA in a saner manner.

> I've just reverted this change.

As long as you can safely call `wait_for_done()` multiple times without blocking forever this should be fine.